src/main/java/jcifs/netbios/NameServiceClientImpl.java

```diff
@@ -0,0 +1,1072 @@
+                if ( response == null || response.received ) {
+                    continue;
+                }
+                synchronized ( response ) {
+                    response.readWireFormat(this.rcv_buf, 0);
+                    response.received = true;
+
+                    if ( log.isTraceEnabled() ) {
+                        log.trace(response.toString());
+                        log.trace(Hexdump.toHexString(this.rcv_buf, 0, this.in.getLength()));
+                    }
+
+                    response.notify();
+                }
+            }
+        }
+        catch ( SocketTimeoutException ste ) {
+            log.trace("Socket timeout", ste);
+        }
+        catch ( Exception ex ) {
+            log.warn("Uncaught exception in NameServiceClient", ex);
+        }
+        finally {
+            tryClose();
+        }
+    }
+
+
+    void send ( NameServicePacket request, NameServicePacket response, int timeout ) throws IOException {
+        Integer nid = null;
+        int max = this.transportContext.getConfig().getWinsServers().length;
+
+        if ( max == 0 ) {
+            max = 1; /* No WINs, try only bcast addr */
+        }
+
+        synchronized ( response ) {
+            while ( max-- > 0 ) {
+                try {
+                    synchronized ( this.LOCK ) {
+                        request.nameTrnId = getNextNameTrnId();
+                        nid = new Integer(request.nameTrnId);
+
+                        this.out.setAddress(request.addr);
+                        this.out.setLength(request.writeWireFormat(this.snd_buf, 0));
+                        response.received = false;
+
+                        this.responseTable.put(nid, response);
+                        ensureOpen(timeout + 1000);
+                        this.socket.send(this.out);
+
+                        if ( log.isTraceEnabled() ) {
+                            log.trace(request.toString());
+                            log.trace(Hexdump.toHexString(this.snd_buf, 0, this.out.getLength()));
+                        }
+                    }
+
+                    long start = System.currentTimeMillis();
+                    while ( timeout > 0 ) {
+                        response.wait(timeout);
+
+                        /*
+                         * JetDirect printer can respond to regular broadcast query
+                         * with node status so we need to check to make sure that
+                         * the record type matches the question type and if not,
+                         * loop around and try again.
+                         */
+                        if ( response.received && request.questionType == response.recordType )
+                            return;
+
+                        response.received = false;
+                        timeout -= System.currentTimeMillis() - start;
+                    }
+
+                }
+                catch ( InterruptedException ie ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)